types/wlr_gamma_control_v1.c

```diff
@@ -0,0 +1,265 @@
+#include <assert.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <wayland-server.h>
+#include <wlr/interfaces/wlr_output.h>
+#include <wlr/types/wlr_gamma_control_v1.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+#include "wlr-gamma-control-unstable-v1-protocol.h"
+
+#define GAMMA_CONTROL_MANAGER_V1_VERSION 1
+
+static void gamma_control_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void gamma_control_destroy(struct wlr_gamma_control_v1 *gamma_control) {
+	if (gamma_control == NULL) {
+		return;
+	}
+	wlr_output_set_gamma(gamma_control->output, 0, NULL, NULL, NULL);
+	wl_resource_set_user_data(gamma_control->resource, NULL);
+	wl_list_remove(&gamma_control->output_destroy_listener.link);
+	wl_list_remove(&gamma_control->link);
+	free(gamma_control);
+}
+
+static void gamma_control_send_failed(
+		struct wlr_gamma_control_v1 *gamma_control) {
+	zwlr_gamma_control_v1_send_failed(gamma_control->resource);
+	gamma_control_destroy(gamma_control);
+}
+
+static const struct zwlr_gamma_control_v1_interface gamma_control_impl;
+
+static struct wlr_gamma_control_v1 *gamma_control_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwlr_gamma_control_v1_interface,
+		&gamma_control_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void gamma_control_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_gamma_control_v1 *gamma_control =
+		gamma_control_from_resource(resource);
+	gamma_control_destroy(gamma_control);
+}
+
+static void gamma_control_handle_output_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_gamma_control_v1 *gamma_control =
+		wl_container_of(listener, gamma_control, output_destroy_listener);
+	gamma_control_destroy(gamma_control);
+}
+
+static void gamma_control_handle_set_gamma(struct wl_client *client,
+		struct wl_resource *gamma_control_resource, int fd) {
+	struct wlr_gamma_control_v1 *gamma_control =
+		gamma_control_from_resource(gamma_control_resource);
+	if (gamma_control == NULL) {
+		goto error_fd;
+	}
+
+	uint32_t ramp_size = wlr_output_get_gamma_size(gamma_control->output);
+	size_t table_size = ramp_size * 3 * sizeof(uint16_t);
+
+	off_t fd_size = lseek(fd, 0, SEEK_END);
+	// Skip checks if kernel does no support seek on buffer
+	if (fd_size != -1 && (size_t)fd_size != table_size) {
+		wl_resource_post_error(gamma_control_resource,
+			ZWLR_GAMMA_CONTROL_V1_ERROR_INVALID_GAMMA,
+			"The gamma ramps don't have the correct size");
+		goto error_fd;
+	}
+	lseek(fd, 0, SEEK_SET);
```

Comment:
I'd just skip the whole seeking because you test whether enough data was read later anyway. Also, why use a different error here than after failed `read()`?

Reply:
This check was originally here because I wasn't setting `O_NONBLOCK`, but I guess now it can be merged with the `read` check. The error is different because a `read` can fail for other reasons than the file being too short (in which case the client did nothing wrong and we shouldn't send a protocol error), but I can handle both cases.